google/cloud/bigtable/data/_metrics/data_model.py

```diff
@@ -0,0 +1,514 @@
+        """
+        if (
+            self.state != OperationState.BETWEEN_ATTEMPTS
+            and self.state != OperationState.CREATED
+        ):
+            return self._handle_error(
+                INVALID_STATE_ERROR.format("start_attempt", self.state)
+            )
+
+        # find backoff value
+        if self.backoff_generator and len(self.completed_attempts) > 0:
+            # find the attempt's backoff by sending attempt number to generator
+            backoff = self.backoff_generator.send(len(self.completed_attempts) - 1)
+        else:
+            backoff = 0
+
+        self.active_attempt = ActiveAttemptMetric(backoff_before_attempt=backoff)
+
+    def add_response_metadata(self, metadata: dict[str, bytes | str]) -> None:
+        """
+        Attach trailing metadata to the active attempt.
+
+        If not called, default values for the metadata will be used.
+
+        If the operation was completed or there is no active attempt,
+        will raise an exception or warning based on the value of ALLOW_METRIC_EXCEPTIONS.
+
+        Args:
+          - metadata: the metadata as extracted from the grpc call
+        """
+        if self.state != OperationState.ACTIVE_ATTEMPT:
+            return self._handle_error(
+                INVALID_STATE_ERROR.format("add_response_metadata", self.state)
+            )
+        if self.cluster_id is None or self.zone is None:
+            # BIGTABLE_METADATA_KEY should give a binary string with cluster_id and zone
+            blob = cast(bytes, metadata.get(BIGTABLE_METADATA_KEY))
+            if blob:
+                parse_result = self._parse_response_metadata_blob(blob)
+                if parse_result is not None:
+                    self.zone, self.cluster_id = parse_result
+                else:
+                    self._handle_error(
+                        f"Failed to decode {BIGTABLE_METADATA_KEY} metadata: {blob!r}"
+                    )
+        # SERVER_TIMING_METADATA_KEY should give a string with the server-latency headers
+        timing_header = cast(str, metadata.get(SERVER_TIMING_METADATA_KEY))
+        if timing_header:
+            timing_data = SERVER_TIMING_REGEX.match(timing_header)
+            if timing_data and self.active_attempt:
+                # convert from milliseconds to seconds
+                self.active_attempt.gfe_latency = float(timing_data.group(1)) / 1000
+
+    @staticmethod
+    @lru_cache(maxsize=32)
+    def _parse_response_metadata_blob(blob: bytes) -> Tuple[str, str] | None:
+        """
+        Parse the response metadata blob and return a dictionary of key-value pairs.
+
+        Function is cached to avoid parsing the same blob multiple times.
+
+        Args:
+          - blob: the metadata blob as extracted from the grpc call
+        Returns:
+          - a tuple of zone and cluster_id, or None if parsing failed
+        """
+        try:
```

Comment:
This should be returned in the ResponseParams proto and you can access zoneId and clusterId from the ResponseParams proto, instead of decode the blob

Reply:
I don't remember seeing this in the response. How is it sent?

Reply:
Oh, I think I see what you mean, construct a proto using the blob?